### ISSUE TITLE
JSON naming conventions

### ISSUE THREAD

**Issue by pmackay** [alphagov/tech-and-data-standards#14]
_Copied from #9:_

**Naming Standards** – adopting camelCase rather than snake_case or hypens. It would be useful to have a common messaging model so that we call the same elements the same things.

Refs:
- https://github.com/18F/api-standards/issues/40
- http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#snake-vs-camel

Also I notice that Registers are using hyphens. @psd was there some specific thinking behind that?

**Comment by rossjones** [comment]
I think the hyphen in a JSON key thing (from registers) might actually be a bug. 
It’ll almost certainly trick people into breaking their own code if they are 
using it from Javascript. We should avoid advocating hyphens in JSON keys.

``` javascript
$.ajax({
    url: "...",
    dataType: "json",
        success: function(obj) {
       // this will complain it can’t find number 
       // will force people to user obj["serial-number"]
       console.log(obj.serial-number)
    }
});
```

There's a workaround, but I'm still pretty sure it'll catch people out.

Edit: I'd report it as a bug, just not sure which openregistry repo to do it on.

**Comment by philandstuff** [comment]
Before saying anything specific about registers, my first concern is that this seems the kind of issue where it would be difficult-to-impossible to gain consensus, and so unlikely to lead to a standard.

**Comment by psd** [comment]
There has been a lot of thinking, and some user research behind using kebab-case: our data is independent of representation, so whilst CamelCase in JSON works well for web development in 2016, it has a shorter longevity than the data we're recording, and looks alien in other representations such as CSV (needed by analysts) TSV (great for shell scripting) XML based languages and RDF. 

So we went for the most human-readable form, rather than PascalCase or camelCase or snake_case. No matter which you pick, it'll look strange in another representation or programming language, but that doesn't prevent data-binding or building a library which transposes the names programmatically, which is another advantage of kebab or snake-case, it's easy to map to and from CamelCase.

Ultimately this is a design decision and matter of taste. We had to pick something, and changing is a big change for us for little to no benefit.

in JavaScript you access the names using object['serial-number']

**Comment by rossjones** [comment]
Thanks for the feedback @psd, much appreciated.  

I was just taking a pragmatic approach to something that will bite people today, regardless of what'll happen in the future.  Obviously you can access the value as a dictionary, but I've seen lots of code where the developers prefer property access. The prevalence of pages for [hyphen in json](https://www.google.co.uk/search?q=json+hyphen) seems to support that.

Given it is _just_ a serialisation, could it be not be fixed in the JSON renderer?  Is it critical that the naming is consistent across all the different serialisations?

Edit: Happy to submit a PR and discuss it there if someone could point me to the right place?

**Comment by pmackay** [comment]
Likewise, all feedback very useful, thanks.

@philandstuff you might be right, it was mentioned originally in #9 and its something I've been debating on the recent work I'm doing. If the answer is "this community cannot or dont want to define a standard" thats a useful thing to know :) 

Another alternative would be gradual consistency by numbers, but that would need some register of APIs (as also discussed yesterday) so the formats used could be reviewed.

**Comment by psd** [comment]
We're not likely to change the field-names in different representations, as that makes generating an objecthash more error-prone.

You could create a JavaScript client to turn them into an object in with camel-case simply enough — though JavaScript developers are comfortable dealing with kebab-case in CSS and markup anyway.

**Comment by rossjones** [comment]
I'm not being very effective at getting my point across, and if I keep trying to get closer to my point I guess I'll just annoy people. Thanks again for clarifying.

**Comment by robmckinnon** [comment]
I started writing a Ruby client for registers at yesterday's workshop. A similiar situation exists if you want to convert register field-names to method names on a Ruby object, which are snake_case.

As @psd suggested it's possible in a programming language specific client to convert JSON keys as appropriate to that language. This is the approach I've taken. I've previously written a library to make [emerging Ruby class definitions from data](https://github.com/robmckinnon/morph#readme) easier. Using this library I had the barebones of a Ruby register client in 30 minutes.

Feels like the register project has the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) in mind - being conservative with what it does, and hoping consumers can be liberal in what they accept.

**Comment by pmackay** [comment]
@robmckinnon if you put your registers Ruby client on github anytime, please do post a link to it. Would be very interested to have a look and a play.

@psd your points above about why openregister uses kebab-case are useful to know. Is it worth capturing any more detail on the user research, etc that could be a reference for other projects? 

Are there other guidelines that would naturally inform this format decision? E.g. if GOV.UK used Article/Periodical formats (as mentioned [here](https://github.com/alphagov/tech-and-data-standards/issues/18#issuecomment-188422100)) does that require/imply using camelCase?

In short, if consensus on a _standard_ is not possible, are more detailed guidelines are useful fallback? (this could apply to other things discussed in this group of course)

**Comment by edent** [comment]
I can't find an Open Standard which defines what camelCase is ;-)

As this is discussion about a convention, I'll close for now.  Feel free to re-open if I've made a mistake.
